File: 40-combination-sum-ii/combination-sum-ii.py
class Solution:
    def comb(self, i: int, sub: List[int], ans: List[List[int]], candidates: List[int], t: int) -> None:
        if t == 0:
            ans.append(list(sub))
            return
        if i == len(candidates) or t < candidates[i]:
            return
        for ind in range(i, len(candidates)):
            if ind > i and candidates[ind - 1] == candidates[ind]:
                continue
            sub.append(candidates[ind])
            self.comb(ind + 1, sub, ans, candidates, t - candidates[ind])
            sub.pop()
    
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        sub = []
        candidates.sort()
        self.comb(0, sub, ans, candidates, target)
        return ans

File: 40-combination-sum-ii/combination-sum-ii.java
class Solution {
    public void comb(int i, List<Integer> sub, List<List<Integer>> ans, int[] candidates, int t) {
        if (t == 0) {
            ans.add(new ArrayList<>(sub));
            return;
        }
        if (i == candidates.length) return;
        if (t < candidates[i]) return;
        for (int ind = i; ind < candidates.length; ind++) {
            if (ind > i && candidates[ind - 1] == candidates[ind]) continue;
            sub.add(candidates[ind]);
            comb(ind + 1, sub, ans, candidates, t - candidates[ind]);
            sub.remove(sub.size() - 1);
        }
        return;
    }

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> sub = new ArrayList<>();
        Arrays.sort(candidates);
        comb(0, sub, ans, candidates, target);
        return ans;
    }
}